text/0000-assoc-default-groups.md

```diff
@@ -0,0 +1,1468 @@
+        default fn delta() -> [Self::Bar; Self::BETA] {
+            // ERROR! we may not rely on not on `Self::BETA`'s value because
+            // `Self::BETA` is a sibling of `Self::gamma` which is not in the
+            // same group and is not an ancestor either.
+            [9usize, 8, 7]
+        }
+
+        // But this is accepted:
+        default fn delta() -> [Self::Bar; 3] {
+            // OK; we can depend on `Self::Bar == usize`.
+            [9, 8, 7]
+        }
+
+        default {
+            // OK; we can still depend on `Self::Bar == usize`.
+            const EPSILON: Self::Bar = 2;
+
+            fn zeta() -> [Self::Bar; Self::Epsilon] {
+                // OK; We can assume the value of `Self::EPSILON` because it
+                // is a sibling in the same group. We may also assume that
+                // `Self::Bar == usize` because it is an ancestor.
+                [42usize, 24]
+            }
+        }
+    }
+}
+
+struct Eta;
+struct Theta;
+struct Iota;
+
+impl Foo for Eta {
+    // We can override `gamma` without overriding anything else because
+    // `gamma` is the sole member of its sub-group. Note in particular
+    // that we don't have to override `alpha`.
+    fn gamma() -> [Self::Bar; 4] {
+        [43, 42, 41, 40]
+    }
+}
+
+impl Bar for Theta {
+    // Since `EPSILON` and `zeta` are in the same group; we must override
+    // them together. However, we still don't have to override anything
+    // in ancestral groups.
+    const EPSILON: Self::Bar = 0;
+
+    fn zeta() -> [Self::Bar; Self::Epsilon] {
+        []
+    }
+}
+
+impl Bar for Iota {
+    // We have overridden `Bar` which is in the root group.
+    // Since all other items are decendants of the same group as `Bar` is in,
+    // they are allowed to depend on what `Bar` is.
+    type Bar = u8;
+
+    ... // Definitions for all the other items elided for brevity.
+}
+```
+
+[clique]: https://en.wikipedia.org/wiki/Clique_(graph_theory)
+
+In graph theory, a set of a vertices, in a graph, for which each distinct pair
+of vertices is connected by a unique edge is said to form a [clique].
+What the snippet above encodes is a tree of such cliques. In other words,
+we can visualize the snippet as:
+
+```
+                            ┏━━━━━━━━━━━━━━━━━┓
+                            ┃ + type Bar      ┃
+              ┏━━━━━━━━━━━━━┃ + fn alpha      ┃━━━━━━━━━━━━━━┓
+              ┃             ┗━━━━━━━━━━━━━━━━━┛              ┃
+              ┃               ┃             ┃                ┃
+              ┃               ┃             ┃                ┃
+              ▼               ▼             ▼                ▼
+┏━━━━━━━━━━━━━━━┓    ┏━━━━━━━━━━━━━┓    ┏━━━━━━━━━━━━━┓    ┏━━━━━━━━━━━━━━━━━┓
+┃ + const Beta  ┃    ┃ + fn gamma  ┃    ┃ + fn delta  ┃    ┃ + const EPSILON ┃
+┗━━━━━━━━━━━━━━━┛    ┗━━━━━━━━━━━━━┛    ┗━━━━━━━━━━━━━┛    ┃ + fn zeta       ┃
+                                                           ┗━━━━━━━━━━━━━━━━━┛
+```
```

Comment:
My main reaction here is that this is quite complicated.  How necessary is all this?  Could all of these cases be handled by having defaults in `impl`s?  Do we even need _groups_ in the `trait`s themselves?